### PR TITLE
chore(build): upgrade github actions to artifacts v4

### PR DESCRIPTION
**Pull request by garysassano**
Upgrading GitHub Actions to [Artifacts v4](https://github.blog/changelog/2023-12-14-github-actions-artifacts-v4-is-now-generally-available/) for performance improvements.

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

**Comment by tsuf239** [comment]
from the [link](https://github.blog/changelog/2023-12-14-github-actions-artifacts-v4-is-now-generally-available/):
```
This version of the artifact actions include performance improvements and new features, however there are also key differences from previous versions that may require updates to your workflows.

- Artifacts will be scoped to a job rather than a workflow. This allows the artifact to become immediately available to [download from the API](https://docs.github.com/rest/actions/artifacts#download-an-artifact) after being uploaded.
- Multiple jobs in the same run cannot upload to the same artifact.
- A single job can upload a maximum of 10 artifacts.
- Artifacts v4 is not cross-compatible with previous versions. For example, an artifact uploaded using v3 cannot be used with actions/download-artifact@v4.
```

@MarkMcCulloh , I am worried about the first two points, do you think it will break something?

File: .github/workflows/tf-aws-test.yml
Comment (MarkMcCulloh) [inline .github/workflows/tf-aws-test.yml:1]: Could you also update action-download-artifact in the mutation workflow to v3?

**Review by MarkMcCulloh — APPROVED** [review]
Thanks @garysassano, and good callout @tsuf239. There's only one part I'm concerned about:

> Artifacts will be scoped to a job rather than a workflow. 

This might be a problem, but I'm not sure. We use https://github.com/dawidd6/action-download-artifact for the mutation job. TBH I'd rather give this a shot and monitor the mutation job. If it fails we'll roll this back for now.

**Comment by staycoolcall911** [comment]
@MarkMcCulloh - do you know why mergify failed?

**Comment by MarkMcCulloh** [comment]
The merge queue PR failed here https://github.com/winglang/wing/actions/runs/7253027111/job/19758792949?pr=5248

Normally I would say it's a transient error, but I'm a little concerned that it failed on this exact step that was updated. I'll update this PR and try it again.

**Pull request merged**